(WIP) Mapping for the Traktor Kontrol Z2

JoergAtGithub opened the pull request:
New HID mapping for the Native Instrument Kontrol Z2 mixer.

Note: This mapping requires features from 2.4 or Main branch and can not work with 2.2 or 2.3 builds!

The Documentation PR is https://github.com/mixxxdj/manual/pull/236
Link to the Preview of the Z2 Mapping-Manual: https://deploy-preview-236--mixxx-manual.netlify.app/hardware/controllers/native_instruments_traktor_kontrol_z2.html

JoergAtGithub committed:
Initial commit of a Mixxx mapping for the Traktor Z2
The code is derivived from this PR: https://github.com/mixxxdj/mixxx/pull/3031

JoergAtGithub committed:
Vinylcontrol buttons
Microphone controls

ywwg commented:
on the s3, I found that the outgoing HID messages have to be exactly the right length for them to work.  I used usblyzer to sniff the outgoing packets from traktor.

JoergAtGithub commented:
I could get nearly all LEDs to work, except the last one (an LED in a 7segment display). I'ts byte 95 of a packet of 95 bytes. I get always the following error message (German windows settings):


Debug [Controller]: HID TraktorZ2: debugLights
Debug [Controller]: HID 2

Debug [Controller]: HID i 0 70,70,70,70,70,00,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,FF,00,00,FF,00,00,00,00,00,00,00,00,34,34,34,00,
00,34,00,FF,00,00,00,34,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,70,
Debug [Controller]: HID 95

Debug [Controller]: HID 128 95

**_Warning [Controller]: Unable to send data to "Traktor Kontrol Z2 13EA_4" serial # "C86013EA" : "Der angegebene Benutzerpuffer ist für den angeforderten Vorgang
nicht zulässig."_**
Debug [Controller]: HID i 1 FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF
Debug [Controller]: HID 41

Debug [Controller]: HID 129 41

Debug [Controller]: 95 bytes sent to Traktor Kontrol Z2 13EA_4 serial # C86013EA (including report ID of 129 )
Debug [Controller]: HID TraktorZ2: Init done!

If I reduce the size to 94 bytes, the other LEDs can be controlled

JoergAtGithub committed:
Could get most LEDs to lit, but the communication fails if I set the correct package size of 95bytes. With 94 bytes it works, but one LED is missing.

ywwg commented on res/controllers/Traktor.Kontrol-Z2-hid-scripts.js:
```diff
@@ -0,0 +1,629 @@
+TraktorZ2.shutdown = function() {
+    // Deactivate all LEDs
+    var dataStrings = [
+        "      00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
```
there is a bug in this code (that was in my code too-- if you have leading spaces, the first item in the list is an empty string so that byte does get set

ywwg replied:
the fix is to eliminate leading spaces in your strings

JoergAtGithub replied:
Thank's for poiting to this! This is the shutdown code, which I've to rework anyway. Currently I'm working on debugLights.

The shutdown code for the Z2 will be tricky anyway, because of the concept of seamless switching between internal and external mixing mode. The idea is:
-At the end of your DJ set, switch one channel from internal to external mixing mode and play an analog vinyl
-DJ1 disconnects his laptop -> Shutdown code must not interfere the music
-DJ2 connects his laptop -> Init code must not interfere the music
-Switch the deck with the analog vinyl back to internal mixing mode

Therefore I will have to decide LED by LED if the shutdown code should reset it, or if the LED represents a still active state.

ywwg commented on res/controllers/Traktor.Kontrol-Z2-hid-scripts.js:
```diff
@@ -0,0 +1,629 @@
+        data[i].length = splitted.length;
+        for (var j = 0; j < splitted.length; j++) {
+            var byteStr = splitted[j];
+            if (byteStr.length === 0) {
```
this line will get activated

ywwg commented on res/controllers/Traktor.Kontrol-Z2-hid-scripts.js:
```diff
@@ -0,0 +1,629 @@
+        HIDDebug(splitted.length + "\n");
+        for (var j = 0; j < splitted.length; j++) {
+            var byteStr = splitted[j];
+            if (byteStr.length === 0) {
```
Add hiddebug here to make sure this isn't happening

ywwg commented:
looks like a low-level HID error message.  I'm not sure why the 95th byte makes it blow up.

JoergAtGithub committed:
Removed LED shutdown code, which doesn't suit the Z2
Corrected LED comments

JoergAtGithub committed:
Basic functionality for Sync button LED outputs

JoergAtGithub committed:
Fixed HID communication error for byte 95 in debugLights. An additional space at the end of the string was the problem

JoergAtGithub committed:
Set the correct USB Product ID for the Z2

JoergAtGithub commented:
Cross-Reference to related Zulip-Chat: https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Traktor.20Kontrol.20Z2

JoergAtGithub committed:
Initial Code for 7 segment loop displays

vbatts commented:
Related work (by the op) for the LEDs on this same board: https://github.com/mixxxdj/mixxx/pull/3051

JoergAtGithub committed:
Changed file names according conventions used by other mappings
Use new command controller.sendFeatureReport from 3051 to get control over the LEDs
Therefore this mapping will not work with 2.3.x
Mapped loop count displays
Mapped VU-Meters
Mapped Hotcue colors
Mapped Macro FX units
Implemented Shift-Functionality
and several small things

JoergAtGithub commented:
This will not go into 2.3, because this mapping can't work without  #3051, which is targeted for 2.4.

JoergAtGithub committed:
Reverted last merge

Holzhaus commented on res/controllers/Traktor Kontrol Z2.hid.xml:
```diff
@@ -0,0 +1,18 @@
+<MixxxControllerPreset mixxxVersion="2.4.0" schemaVersion="1">
+     <info>
+        <name>Traktor Kontrol Z2</name>
+        <author>Jörg Wartenberg (based on the Traktor S3 mapping by Owen Williams)</author>
+        <description>HID Mapping for Traktor Kontrol Z2</description>
```
```suggestion
        <author>Jörg Wartenberg</author>
        <description>HID Mapping for Traktor Kontrol Z2 based on the Traktor S3 mapping by Owen Williams.</description>
```

Holzhaus reviewed (changes requested):
Thank you.

There are code-style issues. Please click the "Details" link for Travis CI below this PR and check the output of the pre-commit job. 

Also, starting with Mixxx 2.3, all controller mappings are now supposed to be documented in the Manual. This is necessary so that other users know how to use this mapping. We also need it to review it properly and check if the mapping makes sense and if the code matches the documentation.

Please fork the mixxxdj/manual repository, check out the manual-2.3.x branch and add documentation to the source/hardware/controllers directory. Open the pull request and add the link to the PR description.

Here's an example how such documentation could look like: https://manual.mixxx.org/2.3/en/hardware/controllers/roland_dj_505.html (If you don't have schematic images and don't want to draw them yourself, just use the same numbering as in the controller's manual and document the buttons like that).

JoergAtGithub committed:
Update res/controllers/Traktor Kontrol Z2.hid.xml

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

JoergAtGithub committed:
Update res/controllers/Traktor Kontrol Z2.hid.xml

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

JoergAtGithub committed:
Fixed codespell failures

JoergAtGithub committed:
Added sort by library column using #3167

JoergAtGithub committed:
Rebase to latest master

JoergAtGithub committed:
Fixed codefactor issues

JoergAtGithub committed:
Fixed codefactor issues

JoergAtGithub committed:
Fixed codefactor and codespell issues

JoergAtGithub committed:
Reduced communication to controller

JoergAtGithub committed:
Fixed bugs:
-Bugs introduced by Search and Replace symbol renaming
-Global variable usage in Timer function usage

JoergAtGithub committed:
Implemented support for fractional beatloops

JoergAtGithub committed:
Added link to new manual page. Page needs to be created.
Removed dead wiki link, which was a placeholder anyway. This page never existed.
Unified header with other NI controller mappings

JoergAtGithub commented on res/controllers/Traktor Kontrol Z2.hid.xml:
```diff
@@ -0,0 +1,18 @@
+        <name>Traktor Kontrol Z2</name>
+        <author>Jörg Wartenberg (based on the Traktor S3 mapping by Owen Williams)</author>
+        <description>HID Mapping for Traktor Kontrol Z2</description>
+        <wiki>http://mixxx.org/wiki/doku.php/traktor_z2</wiki>
```
1.) I removed the link to the non-existing wiki page at all.
2.) I added a link to the manual, and it points to the 2.3 manual, but this PR is rebased to master.

Holzhaus commented:
Pre-commit issues:

```
codespell................................................................Failed
- hook id: codespell
- exit code: 1
res/controllers/Traktor-Kontrol-Z2-hid-scripts.js:829: inbetween ==> between, in between
res/controllers/Traktor-Kontrol-Z2-hid-scripts.js:1155: swiched ==> switched
eslint...................................................................Failed
- hook id: eslint
- exit code: 1
/home/travis/build/mixxxdj/mixxx/res/controllers/Traktor-Kontrol-Z2-hid-scripts.js
  12:1  error  Parsing error: The keyword 'const' is reserved
✖ 1 problem (1 error, 0 warnings)
```

JoergAtGithub committed:
Improved loop display behavior
Fixed codespell errors

JoergAtGithub committed:
Make constants to variables, to fix eslint error

JoergAtGithub committed:
Make constants to variables, to fix eslint error

JoergAtGithub committed:
Fixed codesfactor errors

Tim19801 commented:
Hi, thanks for the mapping. I've tried the setting under Mixxx 2.3 and older versions under Ubuntu. I didn't have luck so far. Mixxx sees the controller and I can indicate which channels I want to use (so it mus recognize my soundcard under the Kontrol Z2 in some sort of way) but it only sends audio over the master output. 

I can't control anything either. Maybe I can also check the template under Windows. 
In the manual it says the Kontrol Z2 is USB compliant and automatically recognized in Linux but I don't have any luck with it. 

Could someone please point me in the right direction? Thanks! Tim

JoergAtGithub commented:
The soundcard is independent from this mapping for Buttons, Knobs, Faders and LEDs.
If you can select Traktor Z2 as Audio source, than the operating system has recognized the soundcard. The soundcard settings should be:
![grafik](https://user-images.githubusercontent.com/64457745/96920556-9a90a300-14ad-11eb-9ab8-5cb21ca68aa5.png)
![grafik](https://user-images.githubusercontent.com/64457745/96920638-b300bd80-14ad-11eb-8f13-f53626a8eba1.png)

JoergAtGithub commented:
Mixxx 2.3.0 is not sufficient for this mapping. To use this mapping, you need a recent build of the Master tree containing the PR #3051 . 

The controller mapping must be selected manually:
![grafik](https://user-images.githubusercontent.com/64457745/96921371-af216b00-14ae-11eb-8d3a-f59f8d151ac7.png)

JoergAtGithub commented:
Please continue general discussions about Mixxx and the Traktor Z2 here: https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Traktor.20Kontrol.20Z2

Tim19801 commented:
> Please continue general discussions about Mixxx and the Traktor Z2 here: https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Traktor.20Kontrol.20Z2

Thank you for the help so far. Sorry for the late responseI think I'll try to get the controller to work in Windows first with the suggested mapping. Will report back on the above link if it also Worked under Linux. I also need to read a tutorial on how to pull the pr in Mixxx. Don't have experience with that so need to sort that out first ;-)

JoergAtGithub committed:
Code style issues

JoergAtGithub committed:
Code style issues

JoergAtGithub committed:
Use new SortByCurrentColumn feature

JoergAtGithub committed:
Fixed code style issues

JoergAtGithub committed:
Fixed code style issues

JoergAtGithub committed:
Fixed code style issues

JoergAtGithub committed:
Get several things to work again after switch to main
-Needed to replace this. keyword by TraktorZ2.
-Changed Shift from 2 state to 3 state
-Mapped Ch1 Quantize to Snap-Button
-Mapped Ch2 Quantize to Quantize-Button
-Implemented initial version of a vinylcontrol output handler
-Added USB status LED

JoergAtGithub committed:
Fixed code style issues

Tim19801 commented:
> You don't need to build Mixxx yourself, to test this. The CI systems build this automatically for each PR.
> To get a Windows download of any PR, click on the continuous-integration/appveyor/pr Details-Link at the bottom of the PR -> Visual Studio 2017... -> Artifacts
> In general, this controller mapping contains nothing platform specific. It should run on Linux, MacOS and Windows. The only important thing is, that you use a recent build of the Main development tree.

Hi Joerg,

Thanks for your time to help me out. Sorry for my late responses. Very busy lately. I'm on Linux right now and not Windows. What I tried to do is clone everything and install it. But it indicates there's no makefile. Can imagine cause I don't think it's complete. I'm absolutely not experienced with GIT. I'll try to sort this out when I'm back on Windows. Or I'll just patiently wait for a new version of Mixxx where the Kontrol Z2 is integrated and will learn a thing or two about GIT in the mean time. I'm not sure I follow your explanation; To get a Windows download of any PR, click on the continuous-integration/appveyor/pr Details-Link at the bottom of the PR -> Visual Studio 2017... -> Artifacts but will check it out when I reinstalled Windows on my laptop. Will let you know if I make progress and have tested this. The Z2 seems a very difficult card on Linux. Didn't have luck with it either on xwax so far and it must be USB compliant. Anyway.. thanks. Tim

Holzhaus commented:
Please open a thread on Zulip chat. We can help you with git and compiling on Linux, but we the PR discussion is not the right place for that.

Tim19801 commented:
Sorry for that. Will pay attention to it from now on.

JoergAtGithub committed:
Added VU meter channel switching ChA/ChC and ChB/ChD and Master Label
Mapped Crossfader Reverse

JoergAtGithub committed:
Refactored VuMeter code

JoergAtGithub committed:
Implemented DeckC / DeckD beat indicator LEDs
Switch 7 segment LED displays off, if track is unloaded

JoergAtGithub committed:
Implemented Filter and EQ Knob Snap
Filter and EQ change now also switches VU-Meter to ChA/B

JoergAtGithub committed:
Implemented Load/Duplicate button LEDs as beat indicator

JoergAtGithub committed:
Shifted headphone buttons assigned to DeckC / DeckD

JoergAtGithub committed:
Code style failures

JoergAtGithub committed:
Code style failures

JoergAtGithub committed:
Code style failures

JoergAtGithub committed:
Added proper switching by Shift between beatjump and beatloop

JoergAtGithub committed:
Shifted Quantize aligns beatgrid now

JoergAtGithub committed:
Show HotCue buttons with Saved-Loops dimmed to 30% brightness

JoergAtGithub committed:
Changed the whole LED control from event based code, to fixed refresh rate HID output, to get the beat indicator LEDs in sync

JoergAtGithub committed:
Code style

JoergAtGithub committed:
Use play instead of play_indicator for logic

JoergAtGithub committed:
Fixed broken VU-Labels

JoergAtGithub committed:
Found out, that the brightness range of all LEDs is 7Bit and not 3Bit

JoergAtGithub committed:
Implemented correct behavior of vinylcontrol/play button REL/INTL
Added prefix k to LED brightness constants

JoergAtGithub committed:
Implemented mapping for single vinyl control status LED on Z2
Fixed deinitialization of USB status LED

JoergAtGithub committed:
Implemented correct initialization at startup (incl. Traktor button state)
Added key display, sync, lock
Made beat display robust
Improved JS code by using local variables where possible

Be-ing commented:
Can you separate the eslint config file changes and make a new PR targeted at main?

JoergAtGithub committed:
Merged in eslint rules for ECMAScript 7
Fixed code formatting

JoergAtGithub committed:
index on Traktor-Z2-2.3: 63d7671440 Merge branch 'main' of https://github.com/mixxxdj/mixxx into Traktor-Z2-2.3

JoergAtGithub committed:
Added a code comments about reports

JoergAtGithub committed:
Added waveform zoom/beatgrid scaling by Snap/Quantize + browse knob

JoergAtGithub committed:
Corrected maximum fader value after discussion at https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Traktor.20Kontrol.20Z2/near/245346743

JoergAtGithub committed:
Mapped button of microphone LED

JoergAtGithub committed:
Adapted Send/GetFeatureReport call to new Uint8Array
Refactured duplicated code for registration of outputs which are identical for multiple decks
Added basic support for mirophone button (work not finished yet)
Use of makeUnbufferedConnection to prevent message overflow
Removed unreliable workarounds for message overflow from mapping code

JoergAtGithub committed:
Removed const qualifier, which breaks the mapping

JoergAtGithub committed:
Try let modifier

JoergAtGithub committed:
Declare TraktorZ2 class in extra line

JoergAtGithub committed:
Replaced CO num_effects, which no longer exist in Mixxx 2.4

JoergAtGithub committed:
Removed remainings of no longer needed HID performance workarounds

JoergAtGithub committed:
Improved comments and minor code formatting

JoergAtGithub committed:
Reimplemented displayLoopCountDigit

JoergAtGithub committed:
Updated to new HID API

JoergAtGithub committed:
Refactured code (Use of classes and ES6 bind)

JoergAtGithub committed:
Added eslint ignore, because TraktorZ2 is not an unused var, it's the instance of the mapping class itself.

JoergAtGithub committed:
Adjusted class membership of timers and hid reports

JoergAtGithub committed:
Code formatting - esp. line breaks

JoergAtGithub committed:
Replaced all calls of HIDDebug by console.log
Suppressed 2 false type warnings

JoergAtGithub committed:
Fixed binding of timers

JoergAtGithub committed:
Mapped Intro/Outro Begin/End, to HotCue pads (activated by SHIFT + Deck-Switch)

Russe commented:
Thanks for implementing this, here's some feedback.

1. I does what I expect. Now I need to practice to get used with it.
2. Last modification date is still January 2022
https://github.com/mixxxdj/mixxx/blob/802a416e68474c347492231c6779ba006f8a8a1c/res/controllers/Traktor-Kontrol-Z2-hid-scripts.js#L4

Russe commented:
There's still one issue with your mapping.
After using your mapping, when restarting Z2 controller without USB the headphone and mic LEDs are in a wrong state.

1. Connect Z2 with USB and start using Mixxx with it
2. Turning off Z2 and Mixxx
3. Disconnect USB from Z2
4. Restart Z2
5. Now headphone LEDs and mic LED are on in bright state. Headphone LEDs and mic LED are bright when function is off, and when activating headphones or mic the LEDs are dimmed.

https://github.com/mixxxdj/mixxx/blob/802a416e68474c347492231c6779ba006f8a8a1c/res/controllers/Traktor-Kontrol-Z2-hid-scripts.js#L768

When changing code from above to code from below everything is working fine again.
`const featureRptF3 = new Uint8Array([0x7F, 0x00]);`

Can you confirm this?

vbatts commented:
Just tried your mappings on main (060b0e9) with my Kontrol Z2. ([my branch](https://github.com/vbatts/mixxx/commits/pr-3038-main))
I see it show up as a controller, but it does not stay "enabled".
Console output shows:
> warning [Controller] Failed. Trying to open with make, model & serial no: 6092 4400 "1DCC46F6"
warning [Controller] Unable to open specific HID device "Traktor Kontrol Z2 46F6_4" Trying now with just make and model. (This may only open the first of multiple identical devices.)
warning [Controller] Unable to open HID device "Traktor Kontrol Z2 46F6_4"
warning [Controller] Failed. Trying to open with make, model & serial no: 6092 4400 "1DCC46F6"
warning [Controller] Unable to open specific HID device "Traktor Kontrol Z2 46F6_4" Trying now with just make and model. (This may only open the first of multiple identical devices.)
warning [Controller] Unable to open HID device "Traktor Kontrol Z2 46F6_4"
warning [Controller] Failed. Trying to open with make, model & serial no: 6092 4400 "1DCC46F6"
warning [Controller] Unable to open specific HID device "Traktor Kontrol Z2 46F6_4" Trying now with just make and model. (This may only open the first of multiple identical devices.)
warning [Controller] Unable to open HID device "Traktor Kontrol Z2 46F6_4"
warning [Controller] Failed. Trying to open with make, model & serial no: 6092 4400 "1DCC46F6"
warning [Controller] Unable to open specific HID device "Traktor Kontrol Z2 46F6_4" Trying now with just make and model. (This may only open the first of multiple identical devices.)
warning [Controller] Unable to open HID device "Traktor Kontrol Z2 46F6_4"

JoergAtGithub commented:
Sounds like a permissions error. This message is written, before the JavaScript-Mapping file is loaded.
Which operating system do you use?

vbatts commented:
debian bullseye

JoergAtGithub commented:
Please ensure that you have set the udev rights correctly: https://raw.githubusercontent.com/mixxxdj/mixxx/main/res/linux/mixxx-usb-uaccess.rules
If this doesn't help, please continue the discussion in the forum or in the Zulip chat, where it's more likely to met a Linux user - and this problem is unrelated to this PR.

vbatts commented:
thanks, and apologies. That got me further. 
It lights up, and I see correspondence between the sliders on the device and mixxx now.
Next I'll experiment on getting the Inputs/Outputs working with the time-encoded vinyls.

JoergAtGithub committed:
-Disabled changing LED brightness for external mixing modes
-Used InputReport bitwise OR impleentation using map command (adapted from Roberts S3 mapping PR)
-Used fast HIDController.fastForIn for some often called loops

JoergAtGithub committed:
Fixed Linter issues

JoergAtGithub committed:
Improved ArrayBuffer / UInt8Array conversion

emsi commented:
When using those mapping files with my Z2 my mixxx crashes dumping core:
```
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Segmentation fault (core dumped)
```
I'm using `2.4-beta-124-g30ee73fb62` on Ubuntu

Swiftb0y commented:
Thank you for reporting your issue. Please understand that in order for us to be able to properly debug this, we need a backtrace. Have a look at the wiki for finding out how to create one: https://github.com/mixxxdj/mixxx/wiki/Creating%20Backtraces

emsi commented:
> Thank you for reporting your issue. Please understand that in order for us to be able to properly debug this, we need a backtrace. Have a look at the wiki for finding out how to create one: https://github.com/mixxxdj/mixxx/wiki/Creating%20Backtraces

You mean just regular backtrace or you want me to send you the core file as well?
```
(gdb) backtrace 
#0  0x0000555555995581 in ?? ()
#1  0x00007ffff6e03f6c in QWidget::event(QEvent*) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#2  0x00007ffff6eb1422 in QFrame::event(QEvent*) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#3  0x00007ffff6dc0713 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#4  0x00007ffff3cafe3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007ffff6dc4c39 in QApplicationPrivate::setFocusWidget(QWidget*, Qt::FocusReason) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#6  0x00007ffff6dfd2d6 in QWidget::setFocus(Qt::FocusReason) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#7  0x00007ffff3cdd41e in QObject::event(QEvent*) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#8  0x00007ffff6dc0713 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#9  0x00007ffff3cafe3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#10 0x00007ffff3cb2f27 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff3d09a67 in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#12 0x00007ffff2b12d3b in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#13 0x00007ffff2b68258 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#14 0x00007ffff2b103e3 in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#15 0x00007ffff3d09112 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00005555557a0bf8 in ?? ()
#17 0x00005555557a7437 in ?? ()
#18 0x00007ffff3ce7793 in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#19 0x00007ffff66b0aea in QOpenGLWindow::resizeEvent(QResizeEvent*) () from /lib/x86_64-linux-gnu/libQt5Gui.so.5
#20 0x00007ffff66745a5 in QWindow::event(QEvent*) () from /lib/x86_64-linux-gnu/libQt5Gui.so.5
#21 0x00005555559d76de in ?? ()
#22 0x00007ffff6dc0713 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt5Widgets.so.5
#23 0x00007ffff3cafe3a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#24 0x00007ffff6669748 in QGuiApplicationPrivate::processExposeEvent(QWindowSystemInterfacePrivate::ExposeEvent*) () from /lib/x86_64-linux-gnu/libQt5Gui.so.5
#25 0x00007ffff663da2c in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Gui.so.5
#26 0x00007fffea72ed6e in ?? () from /lib/x86_64-linux-gnu/libQt5XcbQpa.so.5
#27 0x00007ffff2b12d3b in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#28 0x00007ffff2b68258 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#29 0x00007ffff2b103e3 in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#30 0x00007ffff3d090b8 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#31 0x00007ffff3cae75b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#32 0x00007ffff3cb6cf4 in QCoreApplication::exec() () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#33 0x000055555572d9c0 in ?? ()
#34 0x00005555556ffcfb in ?? ()
#35 0x00007ffff3373d90 in __libc_start_call_main (main=main@entry=0x5555556ff3e0, argc=argc@entry=1, argv=argv@entry=0x7fffffffdb08) at ../sysdeps/nptl/libc_start_call_main.h:58
#36 0x00007ffff3373e40 in __libc_start_main_impl (main=0x5555556ff3e0, argc=1, argv=0x7fffffffdb08, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=0x7fffffffdaf8) at ../csu/libc-start.c:392
#37 0x0000555555720215 in ?? ()
```

emsi commented:
I found a workaround. Just remove the `Traktor_Kontrol_Z2_AD9D_4 1` line from ~/.mixxx.cfg` in the following section:
```
[Controller]
Traktor_Kontrol_Z2_AD9D_4 1
```
Then once mixxx starts just tick enable for the Z2 controller. 
Rinse and repeat after restart. :)

Swiftb0y commented:
> You mean just regular backtrace or you want me to send you the core file as well?

Backtrace is enough. From the backtrace you sent, the crash is probably quite tricky to debug. especially when its only triggered by the controller mapping (and likely only for your environment as well)... 

My best guess is that the controller mapping is being loaded too early (or some other component too late) and the behavior in the controller mapping is causing the crash / race. Manually enabling the controller mapping gets rid of the race... 

The segfault is probably because the `model()` call in `WLibraryTableView::focusInEvent` returns a nullptr. Why that is, no idea? again, probably some race with the library. an extra nullcheck with an early return would work around it, but thats only working around the symptom of a deeper issue...

emsi commented:
Bummer :(
I'll add that my setup is just clean mixxxbetas with two files from this PR copied into `~./mixxx/controllers`. Once the mixxx recognizes Z2 and enable it I just close it without any other changes to trigger the issue.

Luckily I've found a workaround.

JoergAtGithub committed:
Pre-Commit

JoergAtGithub committed:
Updated deprecated COs
Use arrow function for timers

JoergAtGithub committed:
Cleaner handling of the encoders (explicit set to undefined until initialization happens)

JoergAtGithub committed:
Added comment which background info

JoergAtGithub committed:
Code cleanup

JoergAtGithub committed:
Refactored shift state code

JoergAtGithub committed:
Map secondary function of settings button to 2/4 deck switch

Swiftb0y commented:
@JoergAtGithub is this ready for review? Do you want a nitpick review with style/implementation suggestions or just a second look to ensure you didn't make any obvious mistakes?

JoergAtGithub commented:
> @JoergAtGithub is this ready for review? Do you want a nitpick review with style/implementation suggestions or just a second look to ensure you didn't make any obvious mistakes?

Any feedback is appreciated! I need to solve (or workaround) a few glitches before it becomes Ready For Merge, but I do not plan functional changes anymore. So it's a good time to have a look at code style and implementation.

Swiftb0y commented on res/controllers/Traktor-Kontrol-Z2-hid-scripts.js:
```diff
@@ -0,0 +1,2452 @@
+        const sideChannel = ["[Channel1]", "[Channel2]"];
+        const sideOffset = [0, 0];
+        if (this.deckSwitch["[Channel1]"] === 1) {
+            sideChannel[1] = "[Channel1]";
+            sideOffset[1] = 0;  // First 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel1]"] === 2) {
+            sideChannel[1] = "[Channel1]";
+            sideOffset[1] = 4;  // Second 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel1]"] === 3) {
+            sideChannel[1] = "[Channel1]";
+            sideOffset[1] = -1;  // IntroBegin/End and OutroBegin/End mapped to the pads
+        } else if (this.deckSwitch["[Channel3]"] === 1) {
+            sideChannel[1] = "[Channel3]";
+            sideOffset[1] = 0;  // First 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel3]"] === 2) {
+            sideChannel[1] = "[Channel3]";
+            sideOffset[1] = 4;  // Second 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel3]"] === 3) {
+            sideChannel[1] = "[Channel3]";
+            sideOffset[1] = -1;  // IntroBegin/End and OutroBegin/End mapped to the pads
+        }
+        if (this.deckSwitch["[Channel2]"] === 1) {
+            sideChannel[2] = "[Channel2]";
+            sideOffset[2] = 0;  // First 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel2]"] === 2) {
+            sideChannel[2] = "[Channel2]";
+            sideOffset[2] = 4;  // Second 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel2]"] === 3) {
+            sideChannel[2] = "[Channel2]";
+            sideOffset[2] = -1;  // IntroBegin/End and OutroBegin/End mapped to the pads
+        } else if (this.deckSwitch["[Channel4]"] === 1) {
+            sideChannel[2] = "[Channel4]";
+            sideOffset[2] = 0;  // First 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel4]"] === 2) {
+            sideChannel[2] = "[Channel4]";
+            sideOffset[2] = 4;  // Second 4 hotcues mapped to the pads
+        } else if (this.deckSwitch["[Channel4]"] === 3) {
+            sideChannel[2] = "[Channel4]";
+            sideOffset[2] = -1;  // IntroBegin/End and OutroBegin/End mapped to the pads
+        }
```
this is duplicated from `numberButtonHandler`.

Swiftb0y commented on res/controllers/Traktor-Kontrol-Z2-hid-scripts.js:
```diff
@@ -0,0 +1,2452 @@
+                let green = ((colorCode & 0x00FF00) >> 8);
+                let blue = ((colorCode & 0x0000FF));
+                // Scale color up to 100% brightness
+                let brightnessCorrectionFactor;
+                if ((red > green) && (red > blue)) {
+                    brightnessCorrectionFactor = kLedBright / red;
+                } else if ((green > red) && (green > blue)) {
+                    brightnessCorrectionFactor = kLedBright / green;
+                } else if ((blue > red) && (blue > green)) {
+                    brightnessCorrectionFactor = kLedBright / blue;
+                }
```
```suggestion
                const brightnessCorrectionFactor = kLedBright / Math.max(red, green, blue);
```

Swiftb0y reviewed (changes requested):
this is not a thorough look at all yet. There is lots of code that is unnecessarily copy-pasted and duplicated (I highlighted some in the review, but the code is full of them). I'd highly appreciate if you could condense them a little using utility functions or lookup tables.

Russe commented:
With most recent version from https://raw.githubusercontent.com/JoergAtGithub/mixxx/Traktor-Z2-2.3/res/controllers/Traktor-Kontrol-Z2-hid-scripts.js Z2 starts up with dimmed SHIFT button.
Can this be changed to SHIFT button off?
Or do I get the js file from wrong source?

JoergAtGithub committed:
Pre-commit

paulhartleyuk-lab commented:
Hi,

I've got this running really well- my only issue is the browse control does not let me go up and down the library, it goes left to right for one song only,

Any thoughts?

Cheers

Paul

paulhartleyuk-lab commented:
> The library scrolling direction depends on the Shift button state

Thank you!!!!
